### PR TITLE
Speed improvements for MPFA

### PR DESCRIPTION
**Pull request by keileg**
## Proposed changes
This PR contains several improvements that speed up MPFA discretization (will have a positive effect on MPSA as well). There are no changes to code functionality. 

To summarize:
* Inversion of block matrices, at the core of MPXA discretizations, now uses parallelized numba. This leads to significant speedup of discretization.
* In Mpfa, slight reorganizations of the code (mainly, avoiding to compute the same matrix-matrix products over again) lead to further savings. NOTE: Similar changes have already been done in MPSA, but we might want to revisit that code and see if there are more easy pickings.
* In grid partitioning, a bottleneck in preprocessing before calling metis was essentially removed.
* Improved and modernized other parts of the code that gave minor performance enhancements.

To quantify the changes, anecdotally, I see a >50% reduction of discretization time on the 3d flow benchmark with a regular geometry on my system (this is likely dependent on a host of other factors, but the gains are real and substantial). My expectation is that quite a bit of this will carry over to MPSA.

## Types of changes

What types of changes does your code introduce to PorePy?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] General maintenance / improvements of the code
- [ ] Testing (contribution related mainly to testing of existing functionality)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR._
- [ ] I have added tests that prove my fix is effective or that my feature works : NOT RELEVANT
- [x] I have added necessary documentation 
- [x] I have checked that I have not duplicated existing functionality

## Further comments
* A significant part of the remaining MPFA cost (~30%) relates to the vector source discretization. We should consider to make it possible to opt out of this.
* I also tested various ways of further speeding up the block inversion, but got no gains. It seems this code is as efficient as we can get it now, the main cost is in the actual inversion as implemented by lapack, which essentially is irreducible.
* Under testing, I observed that the block inversion can speed up significantly (again >50%, perhaps much more) by using numpy compiled with MKL, rather than the precompiled version from pip (conda install may be much better here, since it is already compiled with MKL).

### CODE REVIEW

File: src/porepy/grids/partition.py [hunk 1]
```diff
@@ -45,7 +45,10 @@ def partition_metis(g: pp.Grid, num_part: int) -> np.ndarray:
     c2c = g.cell_connection_map()
 
     # Convert the cells into the format required by pymetis
-    adjacency_list = [list(c2c.getrow(i).indices) for i in range(c2c.shape[0])]
+    indptr = c2c.indptr
```
Comment (IvarStefansson) [inline src/porepy/grids/partition.py:48]: Would it be wise to do a type check/conversion here to ensure no problems with csr vs csc?

File: src/porepy/numerics/fv/fvutils.py [hunk 2]
```diff
@@ -847,7 +847,7 @@ def _exclude_matrix(self, ids):
             [[0, 1, 0, 0],
               [0, 0, 0, 1]]
         """
-        col = np.argwhere([not it for it in ids])
+        col = np.argwhere(np.logical_not(ids))
```
Comment (IvarStefansson) [inline src/porepy/numerics/fv/fvutils.py:850]: Good!

File: src/porepy/numerics/fv/mpfa.py [hunk 3]
```diff
@@ -921,7 +921,10 @@ def _flux_discretization(
         # unknown in the discretized sytsem to the right hand side.
         # The negative in front of pr_trace_cell comes from the grad_eqs
         rhs_cells = -sps.vstack([nk_cell, -pr_trace_cell, pr_cont_cell])
-        flux = darcy * igrad * rhs_cells
+        # Compute matrix-matrix product, this is used in several steps below
+        darcy_igrad = darcy * igrad
+
+        flux = darcy_igrad * rhs_cells
```
Comment (IvarStefansson) [inline src/porepy/numerics/fv/mpfa.py:927]: Could this decomposition help us in differentiable tpfa? It seems that you've found an improved factorisation, which might be exactly what we want.

File: src/porepy/numerics/fv/mpfa.py [hunk 4]
```diff
@@ -1013,9 +1016,11 @@ def _discretize_vector_source(
                 etc.
             bound_exclusion: Object that can eliminate faces related to boundary
                 conditions.
-            darcy: discretized Darcy's law
-            igrad: inverse gradient
-            nk_grad_all: nK products on a sub-cell level
+            darcy_igrad: Product of the discrete Darcy law with the basis functions
+                (igrad)
+            dist_igrad: Product of
+            nk_grad_all: Product between nK products on a sub-cell level and the basis
```
Comment (IvarStefansson) [inline src/porepy/numerics/fv/mpfa.py:1022]: Should we rename? nk_igrad or nk_igrad_all?

File: src/porepy/numerics/fv/mpfa.py [hunk 5]
```diff
@@ -1013,9 +1016,11 @@ def _discretize_vector_source(
...
+                functions (igrad)
             nk_grad_paired: nK products after pairing
```
Comment (IvarStefansson) [inline src/porepy/numerics/fv/mpfa.py:1024]: Same as above

File: src/porepy/numerics/linalg/matrix_operations.py [hunk 6]
```diff
@@ -608,48 +608,61 @@ def invert_diagonal_blocks_numba(a: sps.spmatrix, size: np.ndarray) -> np.ndarra
...
 
-            It may be possible to restructure the code to further help numba,
-            this has not been investigated.
+            IMPLEMENTATION NOTES BELOW
+
+            The code consists of a loop over the blocks. For each block, a local square
+            matrix is formed, the inverse is computed using numpy (which again will
+            invoke LAPACK), and the inverse is stored in an (raveled) array. The most
+            complex part of the code is the formation of the local matrix: Since the
+            original matrix is be sparse, there may be zero elements in the blocks
```
Comment (IvarStefansson) [inline src/porepy/numerics/linalg/matrix_operations.py:624]:
```suggestion
            original matrix is sparse, there may be zero elements in the blocks
```

**Review by IvarStefansson — COMMENTED** [review]
Looking forward to benefitting from this!

**Comment by keileg** [comment]
I merge this after clarification with @IvarStefansson

**Pull request merged**